presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java

```diff
@@ -16,53 +16,284 @@
+                    projects)
+                    .flatMap(newChild -> manipulateAdjacentWindowNodes(parent, newChild, context));
+        }
+
+        public abstract Optional<PlanNode> manipulateAdjacentWindowNodes(WindowNode parent, WindowNode child, Context context);
+    }
+
+    public static class MergeAdjacentWindowsOverProjects
+            extends ManipulateAdjacentWindowsOverProjects
     {
-        return PATTERN;
+        public MergeAdjacentWindowsOverProjects(int numProjects)
+        {
+            super(numProjects);
+        }
+
+        @Override
+        public Optional<PlanNode> manipulateAdjacentWindowNodes(WindowNode parent, WindowNode child, Context context)
+        {
+            if (!child.getSpecification().equals(parent.getSpecification()) || dependsOn(parent, child)) {
+                return Optional.empty();
+            }
+
+            ImmutableMap.Builder<Symbol, WindowNode.Function> functionsBuilder = ImmutableMap.builder();
+            functionsBuilder.putAll(parent.getWindowFunctions());
+            functionsBuilder.putAll(child.getWindowFunctions());
+
+            WindowNode mergedWindowNode = new WindowNode(
+                    parent.getId(),
+                    child.getSource(),
+                    parent.getSpecification(),
+                    functionsBuilder.build(),
+                    parent.getHashSymbol(),
+                    parent.getPrePartitionedInputs(),
+                    parent.getPreSortedOrderPrefix());
+
+            return Optional.of(
+                    restrictOutputs(context.getIdAllocator(), mergedWindowNode, ImmutableSet.copyOf(parent.getOutputSymbols()))
+                            .orElse(mergedWindowNode));
+        }
     }
 
-    @Override
-    public Optional<PlanNode> apply(WindowNode parent, Captures captures, Context context)
+    public static class SwapAdjacentWindowsBySpecifications
+            extends ManipulateAdjacentWindowsOverProjects
     {
-        WindowNode child = captures.get(CHILD);
+        public SwapAdjacentWindowsBySpecifications(int numProjects)
+        {
+            super(numProjects);
+        }
 
-        if (!child.getSpecification().equals(parent.getSpecification()) || dependsOn(parent, child)) {
-            return Optional.empty();
+        @Override
+        public Optional<PlanNode> manipulateAdjacentWindowNodes(WindowNode parent, WindowNode child, Context context)
+        {
+            if ((compare(parent, child) < 0) && (!dependsOn(parent, child))) {
+                PlanNode transposedWindows = transpose(parent, child);
+                return Optional.of(
+                        restrictOutputs(context.getIdAllocator(), transposedWindows, ImmutableSet.copyOf(parent.getOutputSymbols()))
+                                .orElse(transposedWindows));
+            }
+            else {
+                return Optional.empty();
+            }
+        }
+
+        private int compare(WindowNode o1, WindowNode o2)
```

Comment:
Please extract this as some `Comparator<WindowNode>`

Reply:
How would that help?  There's nothing expecting that interface needed in the code.

In any case, this patch doesn't change any of the actual comparing or merging logic, and I think such changes could be done in a separate PR.

Reply:
Not much. I thought it would be cleaner, better separation of concepts (aka SRP),

Reply:
@kokosing It's not reusable code, so why make the impression it is? This method is meant to be used only here, it's not universal. SQL spec does not define what's the order between window nodes, so we should not make assumptions like that but suggesting WindowNodes are comparable/orderable in broader context.
The ordering is established only in order to achieve mergeability and/or lowering the repartitioning efforts, which are implementation details of Presto.
Also, before this patch we had `compare(WindowNode, WindowNode)` as a separate method, not part of class implementing `Comparator<WindowNode>`. I introduced this method recently and @martint approved it. Why this patch would require changing it?

Reply:
Please, forgive me. Maybe I was not clear enough. I do not require this change. This was just my suggestion. I understand your concerns, though IMO swapping and comparing are two different things so I would separate them. 

Anyway, I will leave it up to you.

![](https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSuPOg_77RYiwLVMc2GBnn3wzoLAITXnsl_YjI0omwjivMD11eM)